lib/readline.js

```diff
@@ -500,6 +518,60 @@ Interface.prototype._insertString = function(c) {
     // a hack to get the line refreshed if it's needed
     this._moveCursor(0);
   }
+
+  // searchHistory.call(this);
+  this._searchHistory();
+};
+
+Interface.prototype._appendSearchResult = function(result) {
+  this.previewResult = result !== '' ?
+    `\u001b[90m${result}\u001b[39m` : '';
+
+  // Cursor to left edge.
+  cursorTo(this.output, 0);
+  clearScreenDown(this.output);
+
+  // Based on the line and match result
+  // write the data
+  if (this.previewResult !== '') {
+    this.output.write(buildReverseSearchPrompt(this.line, this.previewResult));
+    cursorTo(this.output, this.cursor + kReverseSearchPrompt.length - 2);
+  } else if (this.line === '') {
+    this.output.write(buildReverseSearchPrompt());
+    cursorTo(this.output, this.cursor + kReverseSearchPrompt.length - 2);
+  } else {
+    this.output.write(buildReverseSearchPrompt(this.line, ''));
+    cursorTo(this.output, this.cursor + kFailedReverseSearchPrompt.length);
+  }
+
+};
+
+// Search a text in a given array
+const searchFor = (array, text) => {
+  const historySet = new Set(array);
+  const result = [];
+  for (var i = 0; i < [...historySet].length; i++)
+    if (text.trim() !== '' && array[i].includes(text))
+      result.push(array[i]);
+  return result;
+};
```

Comment:
I still think it is best to only look for a single entry on each iteration instead of looking through the whole history up front. The logic should also be pretty straight forward in that case (check for next item in list should be pretty straight forward).

Reply:
If I get this correct, you are suggesting to do this. If the history is like : 

```
['a', 'aa', 'aaa', 'zzz']
```

and user searches `a`, instead of returning the search result like `['a', 'aa', 'aaa']`(and can navigating when keystroke for `cmd+r`, which is what current implementation does) you are suggesting me to do stop at `a` (first match) and then when user does keystroke of  `cmd+r`, display `aa` and so on. Right? I see both approaches are almost same, but the current implementation does the search and keeps all results in memory.  I don't see any downside here. Let me know if I miss anything here.

Reply:
As you already said: the current approach has to do the work for all potential entries up front instead of just doing work for exactly the things the user required. It is also a more straight forward approach as the cache requires not only additional memory but is also a bit more complex code wise (instead of just calling the function to find the next entry we have to first find all entries, then check if there are still entries left in the cache). These things are negligible in this case but they are still overhead.